### PR TITLE
SAK-33934 reduce assignment count when person is removed from course

### PR DESCRIPTION
**Pull request by RyanAFinney**
added check in the count submissions list to get just the count of the submission for the students active in the course

added checking to do the queries in multiple portions if there are more than 1000 people in the course. Oracle can only handle up to 1000 items in the "IN" clause at a time.

### CODE REVIEW

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java [hunk 1]
```diff
@@ -197,6 +199,26 @@ public long countAssignmentSubmissions(String assignmentId, Boolean graded, Bool
             criteria.add(Restrictions.eq("userSubmission", userSubmission));
         }
 
+        if(numSubLists > 0){
```
Comment (ern) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java]: There is a better solution using: org.sakaiproject.hibernate.HibernateCriterionUtils

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java [hunk 2]
```diff
@@ -1567,8 +1567,27 @@ public int countSubmissions(String assignmentReference, Boolean graded) {
...
+            String[] permissions = serverConfigurationService.getStrings("assignment.submitter.remove.permission");
+            if (permissions != null) {
+                for (String permission : permissions) {
+                    allowAddSubmissionUsers.removeAll(securityService.unlockUsers(permission, resourceString));
+                }
+            } else {
+                allowAddSubmissionUsers.removeAll(securityService.unlockUsers(SECURE_ADD_ASSIGNMENT, resourceString));
+            }
+
+            for(User user : allowAddSubmissionUsers){
```
Comment (ern) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/AssignmentServiceImpl.java]: spacing

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java [hunk 3]
```diff
@@ -179,11 +177,15 @@ public void initializeAssignment(Assignment assignment) {
     }
 
     @Override
-    public long countAssignmentSubmissions(String assignmentId, Boolean graded, Boolean hasSubmissionDate, Boolean userSubmission) {
+    public long countAssignmentSubmissions(String assignmentId, Boolean graded, Boolean hasSubmissionDate, Boolean userSubmission, List<String> userIds) {
+        int numSubLists = userIds.size() / 1000; //oracle can only handle lists of 1000 or less.
```
Comment (ern) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java]: Remove this

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java [hunk 4]
```diff
@@ -179,11 +177,15 @@ public void initializeAssignment(Assignment assignment) {
     }
 
     @Override
-    public long countAssignmentSubmissions(String assignmentId, Boolean graded, Boolean hasSubmissionDate, Boolean userSubmission) {
+    public long countAssignmentSubmissions(String assignmentId, Boolean graded, Boolean hasSubmissionDate, Boolean userSubmission, List<String> userIds) {
```
Comment (ern) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java]: A null userIds list should mean it's not added to the query.

**Comment by ern** [comment]
Hi Ryan,

Please follow the existing pattern of if the userIds list is null then it should not be part of the query. This adds to the versatility of the method and reduces the need for similar methods.

**Comment by davidpbauer** [comment]
The userIds list should no longer be part of the query if it is null.

**Pull request merged**